README.md
# **pmemkv**

[![GHA build status](https://github.com/pmem/pmemkv/workflows/pmemkv/badge.svg?branch=master)](https://github.com/pmem/pmemkv/actions)
[![Coverity Scan Build Status](https://scan.coverity.com/projects/18408/badge.svg)](https://scan.coverity.com/projects/pmem-pmemkv)
[![Coverage Status](https://codecov.io/github/pmem/pmemkv/coverage.svg?branch=master)](https://codecov.io/gh/pmem/pmemkv/branch/master)
[![PMEMKV version](https://img.shields.io/github/tag/pmem/pmemkv.svg)](https://github.com/pmem/pmemkv/releases/latest)
[![Packaging status](https://repology.org/badge/tiny-repos/pmemkv.svg)](https://repology.org/project/pmemkv/versions)

`pmemkv` is a local/embedded key-value datastore optimized for persistent memory.
Rather than being tied to a single language or backing implementation, `pmemkv`
provides different options for language bindings and storage engines.

For more information, including **C API** and **C++ API** see: https://pmem.io/pmemkv.
Documentation is available for every branch/release. For most recent always see (**master** branch):
 * [C++ docs](https://pmem.io/pmemkv/master/doxygen/index.html),
 * [C manpage libpmemkv(3)](https://pmem.io/pmemkv/master/manpages/libpmemkv.3.html).

Latest releases can be found on the ["releases" tab](https://github.com/pmem/pmemkv/releases).

Up-to-date, current support/maintenance status of branches/releases is available on
[pmem.io](https://pmem.io/pmemkv/index.html#releases-support-status).

There is also a small helper library `pmemkv_json_config` provided.
See its [manual](doc/libpmemkv_json_config.3.md) for details.

## Table of contents
1. [Installation](#installation)
2. [Language Bindings](#language-bindings)
    - [C/C++ Examples](#cc-examples)
    - [Other Languages](#other-languages)
3. [Storage Engines](#storage-engines)
4. [Benchmarks](#benchmarks)
5. [Contact us](#contact-us)

## Installation

[Installation guide](INSTALLING.md)
provides detailed instructions how to build and install `pmemkv` from sources,
build rpm and deb packages and explains usage of experimental engines and pool sets.

- [Building from Sources](INSTALLING.md#building-from-sources)
- [Installing on Fedora](INSTALLING.md#installing-on-fedora)
- [Installing on Ubuntu](INSTALLING.md#installing-on-ubuntu)
- [Using Experimental Engines](INSTALLING.md#using-experimental-engines)
- [Building Packages](INSTALLING.md#building-packages)
- [Using a Pool Set](INSTALLING.md#using-a-pool-set)

## Language Bindings

`pmemkv` is written in C/C++ and can be used in other languages - Java, Node.js,
Python, and Ruby.

![pmemkv-bindings](doc/architecture.png)

### C/C++ Examples

Examples for C and C++ can be found within this repository in [examples directory](./examples/).

### Other Languages

The above-mentioned bindings are maintained in separate GitHub repositories, but are still kept in sync with the main `pmemkv` distribution.

* **Java** - https://github.com/pmem/pmemkv-java
* **Node.js** - https://github.com/pmem/pmemkv-nodejs
* **Python** - https://github.com/pmem/pmemkv-python
* **Ruby** - https://github.com/pmem/pmemkv-ruby

## Storage Engines

`pmemkv` provides multiple storage engines that share common API, so every engine can be used with
all language bindings and utilities. Engines are loaded by name at runtime.

| Engine Name  | Description | Experimental | Concurrent | Sorted | Persistent |
| ------------ | ----------- | :-------------: | :-----------: | :-------: | :-------: |
| [cmap](doc/libpmemkv.7.md#cmap) | Concurrent hash map | No | Yes | No | Yes |
| [vsmap](doc/libpmemkv.7.md#vsmap) | Volatile sorted hash map | No | No | Yes | No |
| [vcmap](doc/libpmemkv.7.md#vcmap) | Volatile concurrent hash map | No | Yes | No | No |
| [csmap](doc/ENGINES-experimental.md#csmap) | [Concurrent sorted map](https://pmem.io/libpmemobj-cpp/master/doxygen/classpmem_1_1obj_1_1experimental_1_1concurrent__map.html) | Yes | Yes | Yes | Yes |
| [radix](doc/ENGINES-experimental.md#radix) | [Radix tree](https://pmem.io/libpmemobj-cpp/master/doxygen/classpmem_1_1obj_1_1experimental_1_1radix__tree.html) | Yes | No | Yes | Yes |
| [tree3](doc/ENGINES-experimental.md#tree3) | Persistent B+ tree | Yes | No | No | Yes |
| [stree](doc/ENGINES-experimental.md#stree) | Sorted persistent B+ tree | Yes | No | Yes | Yes |
| [robinhood](doc/ENGINES-experimental.md#robinhood) | Persistent hash map with Robin Hood hashing | Yes | Yes | No | Yes |

The production quality engines are described in the [libpmemkv(7)](doc/libpmemkv.7.md#engines) manual
and the experimental ones are described in the [ENGINES-experimental.md](doc/ENGINES-experimental.md) file.

`pmemkv` also provides testing engines, which may be used in unit tests or for benchmarking application overhead

| Engine Name  | Description | Experimental | Concurrent | Sorted | Persistent |
| ------------ | ----------- | :-------------: | :-----------: | :-------: | :-------: |
| [blackhole](doc/libpmemkv.7.md#blackhole) | Accepts everything, returns nothing | No | Yes | No | No |
| [dram_vcmap](doc/ENGINES-testing.md#dram_vcmap) | Volatile concurrent hash map placed entirely on DRAM | Yes | Yes | No | No |

[Contributing a new engine](CONTRIBUTING.md#creating-new-engines) is easy, so feel encouraged!

## Benchmarks

**Experimental** benchmark based on *leveldb*'s [db_bench](https://github.com/google/leveldb/blob/master/benchmarks/db_bench.cc)
to measure pmemkv's performance is available here:
https://github.com/pmem/pmemkv-bench (previously *pmemkv-tools*).

## Contact us
For more information about **pmemkv**, contact Igor Chorążewicz (igor.chorazewicz@intel.com),
Piotr Balcer (piotr.balcer@intel.com) or post on our **#pmem** Slack channel using
[this invite link](https://join.slack.com/t/pmem-io/shared_invite/enQtNzU4MzQ2Mzk3MDQwLWQ1YThmODVmMGFkZWI0YTdhODg4ODVhODdhYjg3NmE4N2ViZGI5NTRmZTBiNDYyOGJjYTIyNmZjYzQxODcwNDg) or [Google group](https://groups.google.com/group/pmem).


ChangeLog
Fri Sep 02 2022 Łukasz Stolarczuk <lukasz.stolarczuk@intel.com>

	* Version 1.4.1

	This is the last bugfix release for pmemkv 1.4 version.
	Maintenance of this version is no longer supported.

	Notable changes:
	- Fix get_all return value in robinhood engine
	- Fix iterator return value for zeroed keys

Mon Aug 08 2022 Łukasz Stolarczuk <lukasz.stolarczuk@intel.com>

	* Version 1.3.1

	This is the last bugfix release for pmemkv 1.3 version.
	Maintenance of this version is no longer supported.

	It contains all bugfixes and updates from previous patch releases
	(1.2.1, 1.1.1, 1.0.3).

Tue Jul 27 2021 Igor Chorążewicz <igor.chorazewicz@intel.com>

	* Version 1.5.0

	New features:
	- Introduce (optional) caching layer in radix engine. It can be enabled
		via config parameters
	- Add new config's flag - 'create_if_missing' to address the missing functionality.
		If set, pmemkv tries to open the pool and creates it if missing.
		It's mutually exclusive with 'create_or_error_if_exists' flag

	Bug fixes:
	- Return correct status for wrong parameters in pmemkv_config_from_json
		(instead of failing on assert)
	- Fix iterator::key() implementation to return the whole key, not only
		the part up to the first null-terminator
	- Return correct status from memkind-based engines on open failure

	Other changes:
	- Add alias 'create_or_error_if_exists' for 'force_create' config's flag and
		deprecate the latter one. The old flag can still be used and it behaves
		with no difference. It's discouraged from using, though
	- Add dram_vcmap engine based on tbb::concurrent_hash_map for benchmarking
		purposes
	- Switch default CMake's build type from "Debug" to "RelWithDebInfo"
		when compiling from sources - it's relevant for a GitHub users
		(who may e.g. run some benchmarks)

Tue Jul 06 2021 Łukasz Stolarczuk <lukasz.stolarczuk@intel.com>

	* Version 1.2.1

	This is the last bugfix release for pmemkv 1.2 version.
	Maintenance of this version is no longer supported.

	Notable changes:
	- fix of missing checks in move assigment operators
	- stree: fix operator== by adding missing const specifier
	- stree: fix missing check in node_iterator operator=
	- json_config: return error code for wrong parameters, instead of asserting them
	- json config: extend error messages

Mon Jun 28 2021 Łukasz Stolarczuk <lukasz.stolarczuk@intel.com>

	* Version 1.1.1

	This is the last bugfix release for pmemkv 1.1 version.
	Maintenance of this version is no longer supported.

	Notable changes:
	- json_config: return error status for improper parameters instead of
		asserting nullptr (and crashing)
	- fix error messages for NOT_FOUND and STOPPED_BY_CB statuses (which are not errors)
		by omitting setting (error) message from the previous status
	- vcmap: fix put() implementation (to be thread-safe)
		and decrease the number of temporary string allocations
	- vcmap: fix value_type for concurrent_hash_map
	- vcmap: fix to work with the latest TBB (2021.1) #894
	- stree: fix operator==
	- clarify and extend documentation

Mon Feb 15 2021 Igor Chorążewicz <igor.chorazewicz@intel.com>

	* Version 1.4

	Features:
	- new experimental engine - robinhood (concurrent, unsorted,
		persistent map)
	- experimental iterators API which allows to directly modify values
		(or parts of values)
	- experimental transaction API which allows grouping `put` and `remove`
		operations into a single atomic action

	Other changes:
	- optimized put() in radix engine

	Major fixes:
	- calling pmemkv_errormsg() now returns empty message for NOT_FOUND and
		STOPPED_BY_CB (instead of a possible previous, unrelated error)
	- fix vcmap OOM error on put (github issue #798)

	Known issues:
	- vcmap (github issue #894)

Tue Oct 06 2020 Szymon Romik <szymon.romik@intel.com>

	* Version 1.0.3

	This is the last bugfix release for pmemkv 1.0 version.
	Maintenance of this version is no longer supported.

	Notable changes:
	- VSmap::put method refactor (to use emplace)
	- add errormsg() to db class

Fri Oct 02 2020 Szymon Romik <szymon.romik@intel.com>

	* Version 1.3

	This release introduces a new experimental engine - radix (single-threaded
	sorted map, backed by libpmemobj-cpp's radix_tree container).

	We have also extended configuration class API and redesigned optimized
	version of stree engine.

	Features:
	- radix engine (single-threaded sorted map)
	- config setters with type safety for common config fields
	- stree engine optimization (single-threaded sorted map with custom
		comparator support)

	Major fixes:
	- fixed operator== in stree engine
	- fixed missing checks in move assignment operators

	Other changes:
	- operator<< for pmem::kv::status

	Known issues:
	- vcmap (github issues #623, #798)

Fri May 29 2020 Szymon Romik <szymon.romik@intel.com>

	* Version 1.2

	This release introduces a new experimental engine - csmap (concurrent
	sorted map, backed by libpmemobj-cpp's concurrent_map).

	We have also provided support for custom comparators for sorted engines.

	Features:
	- csmap engine (concurrent sorted map)
	- support for custom comparator for sorted engines

	Optimizations:
	- vsmap::put() now uses emplace() internally

	Other changes:
	- test framework refactor (scenario-based tests, extednded tests for
		concurrentcy and sorting)

Fri Feb 07 2020 Igor Chorążewicz <igor.chorazewicz@intel.com>

	* Version 1.0.2

	This is a bugfix release for pmemkv 1.0

	This release:

pmemkv.cc
/*
 * Copyright 2017, Intel Corporation
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *
 *     * Neither the name of the copyright holder nor the names of its
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <algorithm>
#include <cstring>
#include <iostream>
#include <list>
#include <unistd.h>
#include "pmemkv.h"

#define DO_LOG 0
#define LOG(msg) if (DO_LOG) std::cout << "[pmemkv:" << GetName() << "] " << msg << "\n"

namespace pmemkv {

KVTree::KVTree(const string& name) : name(name) {
    LOG("Opening");
    if (access(GetNamePtr(), F_OK) != 0) {
        LOG("Creating pool");
        pop_ = pool<KVRoot>::create(GetNamePtr(), "pmemkv",
                                    PMEMOBJ_MIN_POOL * 138,  // todo #6, make configurable
                                    S_IRWXU);
    } else {
        pop_ = pool<KVRoot>::open(GetNamePtr(), "pmemkv");
    }
    Recover();
    LOG("Opened ok");
}

KVTree::~KVTree() {
    LOG("Closing");
    Shutdown();
    pop_.close();
    LOG("Closed ok");
}

// ===============================================================================================
// KEY/VALUE METHODS
// ===============================================================================================

KVStatus KVTree::Delete(const string& key) {
    LOG("Delete key=" << key.c_str());
    auto leafnode = LeafSearch(key);
    if (!leafnode) {
        LOG("   head not present");
        return OK;
    }
    const uint8_t hash = PearsonHash(key.c_str(), key.length());
    for (int slot = NODE_KEYS; slot--;) {
        if (leafnode->hashes[slot] == hash) {
            if (strcmp(leafnode->keys[slot].c_str(), key.c_str()) == 0) {
                LOG("   freeing slot=" << slot);
                leafnode->hashes[slot] = 0;
                transaction::exec_tx(pop_, [&] {
                    leafnode->leaf->hashes[slot] = 0;
                });
                break;  // no duplicate keys allowed
            }
        }
    }
    return OK;
}

KVStatus KVTree::Get(const string& key, string* value) {
    LOG("Get key=" << key.c_str());
    auto leafnode = LeafSearch(key);
    if (!leafnode) {
        LOG("   head not present");
        return NOT_FOUND;
    }
    const uint8_t hash = PearsonHash(key.c_str(), key.length());
    for (int slot = NODE_KEYS; slot--;) {
        if (leafnode->hashes[slot] == hash) {
            if (strcmp(leafnode->keys[slot].c_str(), key.c_str()) == 0) {
                value->append(leafnode->leaf->values[slot].get_ro().data());
                LOG("   found value=" << *value << ", slot=" << slot);
                return OK;
            }
        }
    }
    LOG("   could not find key");
    return NOT_FOUND;
}

vector<KVStatus> KVTree::MultiGet(const vector<string>& keys, vector<string>* values) {
    LOG("MultiGet for " << keys.size() << " keys");
    vector<KVStatus> status = vector<KVStatus>();
    for (auto& key: keys) {
        string value;
        KVStatus s = Get(key, &value);
        status.push_back(s);
        values->push_back(s == OK ? value : "");
    }
    LOG("MultiGet done for " << keys.size() << " keys");
    return status;
}

KVStatus KVTree::Put(const string& key, const string& value) {
    LOG("Put key=" << key.c_str() << ", value=" << value.c_str());
    const uint8_t hash = PearsonHash(key.c_str(), key.length());
    auto leafnode = LeafSearch(key);
    if (!leafnode) {
        LOG("   adding head leaf");
        leafnode = new KVLeafNode();
        leafnode->is_leaf = true;
        persistent_ptr<KVLeaf> new_leaf;
        auto root = pop_.get_root();
        auto old_head = root->head;
        transaction::exec_tx(pop_, [&] {
            new_leaf = make_persistent<KVLeaf>();
            new_leaf->next = old_head;
            leafnode->leaf = new_leaf;
            LeafFillSpecificSlot(leafnode, hash, key, value, 0);
            root->head = new_leaf;
        });
        top_ = leafnode;
    } else if (LeafFillSlotForKey(leafnode, hash, key, value)) {
        // nothing else to do
    } else {
        LeafSplit(leafnode, hash, key, value);
    }
    return OK;
}

// ===============================================================================================
// PROTECTED LEAF METHODS
// ===============================================================================================

KVLeafNode* KVTree::LeafSearch(const string& key) {
    KVNode* node = top_;
    if (node == nullptr) return nullptr;
    bool matched;
    while (!node->is_leaf) {
        matched = false;
        KVInnerNode* inner = (KVInnerNode*) node;
        const uint8_t keycount = inner->keycount;
        for (uint8_t idx = 0; idx < keycount; idx++) {
            node = inner->children[idx];
            if (strcmp(key.c_str(), inner->keys[idx].c_str()) <= 0) {
                matched = true;
                break;
            }
        }
        if (!matched) node = inner->children[keycount];
    }
    return (KVLeafNode*) node;
}

void KVTree::LeafFillFirstEmptySlot(KVLeafNode* leafnode, const uint8_t hash,
                                    const string& key, const string& value) {
    for (int slot = NODE_KEYS; slot--;) {
        if (leafnode->hashes[slot] == 0) {
            LeafFillSpecificSlot(leafnode, hash, key, value, slot);
            return;
        }
    }
}

bool KVTree::LeafFillSlotForKey(KVLeafNode* leafnode, const uint8_t hash,
                                const string& key, const string& value) {
    // scan for empty/matching slots
    int last_empty_slot = -1;
    int key_match_slot = -1;
    for (int slot = NODE_KEYS; slot--;) {
        auto slot_hash = leafnode->hashes[slot];
        if (slot_hash == 0) {
            last_empty_slot = slot;
        } else if (slot_hash == hash) {
            if (strcmp(leafnode->keys[slot].c_str(), key.c_str()) == 0) {
                key_match_slot = slot;
                break;  // no duplicate keys allowed
            }
        }
    }

    // update suitable slot if found
    int slot = key_match_slot >= 0 ? key_match_slot : last_empty_slot;
    if (slot >= 0) {
        LOG("   filling slot=" << slot);
        transaction::exec_tx(pop_, [&] {
            LeafFillSpecificSlot(leafnode, hash, key, value, slot);
        });
    }
    return slot >= 0;
}

void KVTree::LeafFillSpecificSlot(KVLeafNode* leafnode, const uint8_t hash,
                                  const string& key, const string& value, const int slot) {
    auto leaf = leafnode->leaf;
    if (leafnode->hashes[slot] == 0) {
        leaf->keys[slot].get_rw().set(key.c_str());
        leafnode->keys[slot] = key;
    }
    leafnode->hashes[slot] = hash;
    leaf->hashes[slot] = hash;
    leaf->values[slot].get_rw().set(value.c_str());
}

void KVTree::LeafSplit(KVLeafNode* leafnode, const uint8_t hash,
                       const string& key, const string& value) {
    string keys[NODE_KEYS + 1];
    keys[NODE_KEYS] = key;
    for (int slot = NODE_KEYS; slot--;) keys[slot] = leafnode->keys[slot];
    std::sort(std::begin(keys), std::end(keys), [](const string& lhs, const string& rhs) {
        return (strcmp(lhs.c_str(), rhs.c_str()) < 0);
    });
    string split_key = keys[NODE_KEYS_MIDPOINT];
    LOG("   splitting leaf at key=" << split_key);

    // split leaf into two leaves, moving slots that sort above split key to new leaf
    auto new_leafnode = new KVLeafNode();
    new_leafnode->parent = leafnode->parent;
    new_leafnode->is_leaf = true;
    persistent_ptr<KVLeaf> new_leaf;
    auto root = pop_.get_root();
    auto old_head = root->head;
    transaction::exec_tx(pop_, [&] {
        new_leaf = make_persistent<KVLeaf>();
        new_leaf->next = old_head;
        new_leafnode->leaf = new_leaf;
        const auto leaf = leafnode->leaf;
        for (int slot = NODE_KEYS; slot--;) {
            if (strcmp(leafnode->keys[slot].c_str(), split_key.data()) > 0) {
                const KVString slot_key = leaf->keys[slot].get_ro();
                if (slot_key.is_short()) {
                    new_leaf->keys[slot].get_rw().set_short(slot_key.data());
                } else new_leaf->keys[slot].swap(leaf->keys[slot]);
                const KVString slot_value = leaf->values[slot].get_ro();
                if (slot_value.is_short()) {
                    new_leaf->values[slot].get_rw().set_short(slot_value.data());
                } else new_leaf->values[slot].swap(leaf->values[slot]);
                new_leafnode->hashes[slot] = leafnode->hashes[slot];
                new_leafnode->keys[slot] = leafnode->keys[slot];
                leafnode->keys[slot].clear();
                new_leaf->hashes[slot] = leafnode->hashes[slot];
                leafnode->hashes[slot] = 0;
                leaf->hashes[slot] = 0;
            }
        }
        auto target = strcmp(key.c_str(), split_key.data()) > 0 ? new_leafnode : leafnode;
        LeafFillFirstEmptySlot(target, hash, key, value);
        root->head = new_leaf;
    });

    // recursively update volatile parents outside persistent transaction
    LeafUpdateParentsAfterSplit(leafnode, new_leafnode, &split_key);
}

void KVTree::LeafUpdateParentsAfterSplit(KVNode* node, KVNode* new_node,
                                         string* split_key) {
    if (!node->parent) {
        LOG("   creating new top node for split_key=" << *split_key);
        auto top = new KVInnerNode();
        top->keycount = 1;
        top->keys[0] = *split_key;
        top->children[0] = node;
        top->children[1] = new_node;
        node->parent = top;
        new_node->parent = top;
        top_ = top;                                                      // assign new top node
        return;                                                          // end recursion
    }

    LOG("   updating parents for split_key=" << *split_key);
    KVInnerNode* inner = (KVInnerNode*) node->parent;
    { // insert split_key and new_node into inner node in sorted order
        const uint8_t keycount = inner->keycount;
        int idx = 0;  // position where split_key should be inserted
        while (idx < keycount && inner->keys[idx].compare(*split_key) <= 0) idx++;
        for (int i = keycount - 1; i >= idx; i--) inner->keys[i + 1] = inner->keys[i];
        for (int i = keycount; i >= idx; i--) inner->children[i + 1] = inner->children[i];
        inner->keys[idx] = *split_key;
        inner->children[idx + 1] = new_node;
        inner->keycount = (uint8_t) (keycount + 1);
    }
    const uint8_t keycount = inner->keycount;
    if (keycount <= INNER_KEYS) return;                                  // end recursion

    // split inner node at the midpoint, update parents as needed
    auto new_inner = new KVInnerNode();                                  // allocate new node
    new_inner->parent = inner->parent;                                   // set parent reference
    for (int i = INNER_KEYS_UPPER; i < keycount; i++) {                  // copy all upper keys
        new_inner->keys[i - INNER_KEYS_UPPER] = inner->keys[i];          // copy key string
    }
    for (int i = INNER_KEYS_UPPER; i < keycount + 1; i++) {              // copy all upper children
        new_inner->children[i - INNER_KEYS_UPPER] = inner->children[i];  // copy child reference
        new_inner->children[i - INNER_KEYS_UPPER]->parent = new_inner;   // set parent reference
    }
    new_inner->keycount = INNER_KEYS_MIDPOINT;                           // always half the keys
    string new_split_key = inner->keys[INNER_KEYS_MIDPOINT];             // save for recursion
    inner->keycount = INNER_KEYS_MIDPOINT;                               // half of keys remain
    LeafUpdateParentsAfterSplit(inner, new_inner, &new_split_key);       // recursive update
}

// ===============================================================================================
// PROTECTED LIFECYCLE METHODS
// ===============================================================================================

void KVTree::Recover() {
    LOG("Recovering");
    auto root = pop_.get_root();
    if (!root->head) {
        LOG("   creating root");
        transaction::exec_tx(pop_, [&] {
            root->opened = 1;
            root->closed = 0;
        });
    } else {
        LOG("   recovering head: opened=" << root->opened << ", closed=" << root->closed);
        // todo #11, are opened/closed counts necessary?
        RebuildNodes();
        transaction::exec_tx(pop_, [&] { root->opened = root->opened + 1; });
    }
    LOG("Recovered ok");
}

void KVTree::RebuildNodes() {
    LOG("   rebuilding nodes");

    // traverse persistent leaves to build list of leaves to recover
    std::list<KVRecoveredLeaf*> leaves;
    auto leaf = pop_.get_root()->head;
    while (leaf != nullptr) {
        auto leafnode = new KVLeafNode();
        leafnode->leaf = leaf;
        leafnode->is_leaf = true;

        // find highest sorting key in leaf, while recovering all hashes
        char* max_key = nullptr;
        for (int slot = NODE_KEYS; slot--;) {
            leafnode->hashes[slot] = leaf->hashes[slot];
            if (leafnode->hashes[slot] == 0) continue;
            char* key = leaf->keys[slot].get_ro().data();
            if (max_key == nullptr || strcmp(max_key, key) < 0) max_key = key;
            leafnode->keys[slot] = key;
        }

        // use highest sorting key to decide how to recover the leaf
        if (max_key == nullptr) {
            // todo #10, handle empty leaf nodes
        } else {
            auto rleaf = new KVRecoveredLeaf;
            rleaf->leafnode = leafnode;
            rleaf->max_key = max_key;
            leaves.push_back(rleaf);
        }

        leaf = leaf->next ? leaf->next : nullptr;  // advance to next linked leaf
    }

    // sort recovered leaves in ascending key order
    leaves.sort([](const KVRecoveredLeaf* lhs, const KVRecoveredLeaf* rhs) {
        return (strcmp(lhs->max_key, rhs->max_key) < 0);
    });

    // reconstruct top/inner nodes using adjacent pairs of recovered leaves
    top_ = nullptr;
    while (!leaves.empty()) {
        KVRecoveredLeaf* rleaf = leaves.front();
        KVLeafNode* leafnode = rleaf->leafnode;
        if (top_ == nullptr) top_ = leafnode;
        leaves.pop_front();
        if (!leaves.empty()) {
            string split_key = string(rleaf->max_key);
            KVLeafNode* nextnode = leaves.front()->leafnode;
            nextnode->parent = leafnode->parent;
            LeafUpdateParentsAfterSplit(leafnode, nextnode, &split_key);
        }
        delete rleaf;
    }

    LOG("   rebuilt nodes ok");
}

void KVTree::Shutdown() {
    LOG("Shutting down");
    auto root = pop_.get_root();
    transaction::exec_tx(pop_, [&] { root->closed = root->closed + 1; });
    LOG("Shut down ok");
}

// ===============================================================================================
// PEARSON HASH METHODS
// ===============================================================================================

// Pearson hashing lookup table from RFC 3074
const uint8_t PEARSON_LOOKUP_TABLE[256] = {
        251, 175, 119, 215, 81, 14, 79, 191, 103, 49, 181, 143, 186, 157, 0,
        232, 31, 32, 55, 60, 152, 58, 17, 237, 174, 70, 160, 144, 220, 90, 57,
        223, 59, 3, 18, 140, 111, 166, 203, 196, 134, 243, 124, 95, 222, 179,
        197, 65, 180, 48, 36, 15, 107, 46, 233, 130, 165, 30, 123, 161, 209, 23,
        97, 16, 40, 91, 219, 61, 100, 10, 210, 109, 250, 127, 22, 138, 29, 108,
        244, 67, 207, 9, 178, 204, 74, 98, 126, 249, 167, 116, 34, 77, 193,
        200, 121, 5, 20, 113, 71, 35, 128, 13, 182, 94, 25, 226, 227, 199, 75,
        27, 41, 245, 230, 224, 43, 225, 177, 26, 155, 150, 212, 142, 218, 115,
        241, 73, 88, 105, 39, 114, 62, 255, 192, 201, 145, 214, 168, 158, 221,
        148, 154, 122, 12, 84, 82, 163, 44, 139, 228, 236, 205, 242, 217, 11,
        187, 146, 159, 64, 86, 239, 195, 42, 106, 198, 118, 112, 184, 172, 87,
        2, 173, 117, 176, 229, 247, 253, 137, 185, 99, 164, 102, 147, 45, 66,
        231, 52, 141, 211, 194, 206, 246, 238, 56, 110, 78, 248, 63, 240, 189,
        93, 92, 51, 53, 183, 19, 171, 72, 50, 33, 104, 101, 69, 8, 252, 83, 120,
        76, 135, 85, 54, 202, 125, 188, 213, 96, 235, 136, 208, 162, 129, 190,
        132, 156, 38, 47, 1, 7, 254, 24, 4, 216, 131, 89, 21, 28, 133, 37, 153,
        149, 80, 170, 68, 6, 169, 234, 151
};

// Modified Pearson hashing algorithm from RFC 3074
uint8_t KVTree::PearsonHash(const char* data, const size_t size) {
    uint8_t hash = (uint8_t) size;
    for (size_t i = size; i > 0;) {
        hash = PEARSON_LOOKUP_TABLE[hash ^ data[--i]];
    }
    // MODIFICATION START
    return (hash == 0) ? (uint8_t) 1 : hash;                             // 0 reserved for "null"
    // MODIFICATION END
}

// ===============================================================================================
// STRING CLASS METHODS
// ===============================================================================================

char* KVString::data() const {
    return str ? str.get() : const_cast<char*>(sso);                     // return short or long
}

void KVString::set(const char* value) {
    size_t value_len = strlen(value);
    if (value_len <= SSO_CHARS) {                                        // setting short value?
        set_short(value);
    } else {                                                             // setting long value?
        if (str) delete_persistent<char[]>(str, strlen(str.get()) + 1);  // free value if present
        str = make_persistent<char[]>(value_len + 1);                    // allocate value pmem
        strcpy(str.get(), value);                                        // copy value data
    }
}

void KVString::set_short(const char* value) {
    if (str) {                                                           // value already present?
        delete_persistent<char[]>(str, strlen(str.get()) + 1);           // free value memory
        str = nullptr;                                                   // zero out pointer
    }
    pmemobj_tx_add_range_direct(sso, SSO_SIZE);                          // add sso buffer to txn
    strcpy(sso, value);                                                  // copy value data
}

} // namespace pmemkv
